File: src/util/java/me/sizableshrimp/adventofcode2025/helper/MathUtil.java
/*
 * AdventOfCode2025
 * Copyright (C) 2025 SizableShrimp
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package me.sizableshrimp.adventofcode2025.helper;

public class MathUtil {
    public static int gcd(int a, int b) {
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }

        return a;
    }

    public static int gcd(int... numbers) {
        int result = numbers[0];

        for (int i = 1; i < numbers.length; i++) {
            result = gcd(result, numbers[i]);
        }

        return result;
    }

    public static long gcd(long a, long b) {
        while (b != 0) {
            long temp = b;
            b = a % b;
            a = temp;
        }

        return a;
    }

    public static long gcd(long... numbers) {
        long result = numbers[0];

        for (int i = 1; i < numbers.length; i++) {
            result = gcd(result, numbers[i]);
        }

        return result;
    }

    public static int lcm(int a, int b) {
        return Math.abs(a * b) / gcd(a, b);
    }

    public static long lcm(int... numbers) {
        long result = numbers[0];

        for (int i = 1; i < numbers.length; i++) {
            result = lcm(result, numbers[i]);
        }

        return result;
    }

    public static long lcm(long a, long b) {
        return Math.abs(a * b) / gcd(a, b);
    }

    public static long lcm(long... numbers) {
        long result = numbers[0];

        for (int i = 1; i < numbers.length; i++) {
            result = lcm(result, numbers[i]);
        }

        return result;
    }
}

File: src/util/kotlin/me/sizableshrimp/adventofcode2025/util/MathUtil.kt
/*
 * AdventOfCode2025
 * Copyright (C) 2025 SizableShrimp
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package me.sizableshrimp.adventofcode2025.util

import me.sizableshrimp.adventofcode2025.helper.MathUtil
import me.sizableshrimp.adventofcode2025.templates.Fraction

infix fun Int.lcm(other: Int) = MathUtil.lcm(this, other)

fun IntArray.lcm() = MathUtil.lcm(*this)

@JvmName("lcmIntCollection")
fun Collection<Int>.lcm() = this.toIntArray().lcm()

infix fun Long.lcm(other: Long) = MathUtil.lcm(this, other)

fun LongArray.lcm() = MathUtil.lcm(*this)

@JvmName("lcmLongCollection")
fun Collection<Long>.lcm() = this.toLongArray().lcm()

infix fun Int.gcd(other: Int) = MathUtil.gcd(this, other)

fun IntArray.gcd() = MathUtil.gcd(*this)

fun Collection<Int>.gcd() = this.toIntArray().gcd()

infix fun Long.gcd(other: Long) = MathUtil.gcd(this, other)

fun LongArray.gcd() = MathUtil.gcd(*this)

fun Collection<Long>.gcd() = this.toLongArray().gcd()

operator fun Fraction.plus(other: Fraction): Fraction = this.plus(other)

operator fun Fraction.minus(other: Fraction): Fraction = this.minus(other)

operator fun Fraction.times(other: Fraction): Fraction = this.times(other)

operator fun Fraction.div(other: Fraction): Fraction = this.divide(other)